core/corehttp/hostname.go

```diff
@@ -0,0 +1,301 @@
+package corehttp
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"regexp"
+	"strings"
+
+	cid "github.com/ipfs/go-cid"
+	core "github.com/ipfs/go-ipfs/core"
+	namesys "github.com/ipfs/go-ipfs/namesys"
+
+	config "github.com/ipfs/go-ipfs-config"
+	nsopts "github.com/ipfs/interface-go-ipfs-core/options/namesys"
+	isd "github.com/jbenet/go-is-domain"
+)
+
+var pathGatewaySpec = config.GatewaySpec{
+	Paths:         []string{"/ipfs/", "/ipns/", "/api/", "/p2p/", "/version"},
+	UseSubdomains: false,
+}
+
+var subdomainGatewaySpec = config.GatewaySpec{
+	Paths:         []string{"/ipfs/", "/ipns/", "/api/", "/p2p/"},
+	UseSubdomains: true,
+}
+
+var defaultKnownGateways = map[string]config.GatewaySpec{
+	// Note: local IPs act as catch-all, no need to define them here
+	//"127.0.0.1":       pathGatewaySpec,
+	//"::1":             pathGatewaySpec,
+	"localhost":       subdomainGatewaySpec,
+	"ipfs.io":         pathGatewaySpec,
+	"gateway.ipfs.io": pathGatewaySpec,
+	"dweb.link":       subdomainGatewaySpec,
+}
+
+// Find content identifier, protocol, and remaining hostname (host+optional port)
+// of a subdomain gateway (eg. *.ipfs.foo.bar.co.uk)
+var subdomainGatewayRegex = regexp.MustCompile("^(.+).(ipfs|ipns|ipld|p2p|api).([^/?#&]+)$")
```

Comment:
```suggestion
var subdomainGatewayRegex = regexp.MustCompile("^(.+)\\.(ipfs|ipns|ipld|p2p|api)\\.([^/?#&]+)$")
```

1. Let's write some tests for this regex.
2. Do we really need a regex? Or can we just split on the first two dots: `CID.PREFIX.EVERYTHINGELSE`?

Reply:
1. Added tests in 077425eb7ea1e0da45fe9fa53b2b16e279512eef
2. Splitting is not enough because we want to support loading DNSLink websites from  localhost, and we don't know how many sub-levels it will produce
   Example: http://en.wikipedia-on-ipfs.org.ipns.localhost:8080

Reply:
Hm. Won't this fail on `http://dist.ipfs.io.ipns.dweb.link`? We may need to some form of suffix lookup:

1. Lookup "link" in the known gateways.
2. Lookup "dweb.link" in the known gateways.

etc.

Reply:
No, iiuc regex in go is greedy by default so it will match all until `.ipns.`
Added test just to be safe tho: c7bd79c104553070cf2de2e39efc5ea76dabb501
(gonna get some sleep, will finish this PR tomorrow)

Reply:
I guess that works. But it does mean we can't have `ipfs.ipfs.io`, or anything like that. I'd prefer to do the manual walk through the known gateways as a more reliable test.

Reply:
I removed `subdomainGatewayRegex` and added test for `.ipfs.*.ipfs.` edge case (ceb111771)

We now do the lookup the way you suggested.